File: 09_DataStructure_Algorithm/0518_day51/02_Combinations/Combinations.java
class Solution {
    private List<List<Integer>> answer = new ArrayList<>();

    public List<List<Integer>> combine(int n, int k) {
        dfs(1, new ArrayList<>(), n, k);
        return answer;
    }

    private void dfs(int start, List<Integer> path, int n, int k) {
        if (path.size() == k) {
            answer.add(new ArrayList<>(path));
            return;
        }

        for (int i = start; i <= n; i++) {
            path.add(i);
            dfs(i + 1, path, n, k);
            path.remove(path.size() - 1);
        }
    }
}

File: 09_DataStructure_Algorithm/0518_day51/02_Combinations/Combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        answer = []

        def combinations(cur, start):
            if len(cur) == k:
                answer.append(cur[:])
                return

            for num in range(start, n + 1):
                cur.append(num)
                combinations(cur, num + 1)
                cur.pop()

        combinations([], 1)

        return answer
